cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/io/cdap/cdap/etl/common/DefaultStageConfigurer.java

```diff
@@ -81,6 +91,37 @@ public void setErrorSchema(@Nullable Schema errorSchema) {
     errorSchemaSet = true;
   }
 
+  @Override
+  public ValidationFailure addFailure(String message, @Nullable String correctiveAction) {
+    ValidationFailure failure = new ValidationFailure(message, correctiveAction);
+    failures.add(failure);
+    return failure;
+  }
+
+  @Override
+  public void throwIfFailure() throws ValidationException {
+    if (!failures.isEmpty()) {
+      for (ValidationFailure failure : failures) {
+        List<ValidationFailure.Cause> causes = failure.getCauses();
+        if (causes.isEmpty()) {
+          causes.add(new ValidationFailure.Cause().with(STAGE, stageName));
+          continue;
+        }
+        for (ValidationFailure.Cause cause : causes) {
+          // stage name is added by the configurer before throwing the validation exception
+          cause.getAttributes().put(STAGE, stageName);
```

Comment:
I would prefer to have a method `cause.addAttribute(key, value)`

Reply:
done